docs/HACKING.md

```diff
@@ -143,3 +143,86 @@ $ make vmoverlay SKIP_INSTALL=1
 
 Of course, you can use this pattern for not just ostree but whatever else you'd
 like to install into the VM (e.g. bubblewrap, libsolv, etc...).
+
+## Using GDB with the rpm-ostree daemon
+
+If you're new to rpm-ostree, before using GDB, it may be helpful to review the
+[daemon architecture doc](../architecture-daemon.md) for an architecture recap.
+
+### Server-side (composes)
+
+Server-side composes do not use the daemon architecture and so one can naturally
+do e.g. `gdb --args rpm-ostree compose tree ...`. If using coreos-assembler, you
+can set the `COSA_RPMOSTREE_GDB` environment variable like this:
+
+```sh
+$ COSA_RPMOSTREE_GDB="gdb --args" cosa build
+```
+
+When cosa gets to the point of invoking rpm-ostree for the compose, it will call
+GDB instead.
+
+### Client-side
+
+On the client side, you need to use the `make vmsync` flow before using GDB
+because it also copies over the source files into `/root/sync`.
+
+For throwaway/fresh VMs, a simple approach is to just layer it using e.g.
+`rpm-ostree install gdb-minimal -A`, and then use it directly. (XXX:
+`apply-live` currently isn't compatible with `make vmsync`, so you'll want to
+reboot for now: https://github.com/ostreedev/ostree/issues/2369).
+
+It's also possible to use GDB from a privileged container. Make sure to use the
+`--pid=host` flag when using e.g. `podman run` so that you can attach to
+processes running on the host. For example:
+
+```
+(host) podman run -ti --privileged --pid=host -v /:/host --name gdb \
```

Comment:
I think this would be better as `toolbox` and we should document this use case in the main docs.

Reply:
Sure, open to it. If you want, feel free to tweak these docs in a follow-up! (Personally still haven't switched over to toolbox because it's missing a few things I still need from my pet container setup. I need to revisit that at some point.)